src/XMTP.types.ts

```diff
@@ -74,30 +78,99 @@ export type PreparedLocalMessage = {
   messageId: string;
   preparedFileUri: `file://${string}`;
   preparedAt: number; // timestamp in milliseconds
-}
+};
 
 // This contains the contents of a message.
 // Each of these corresponds to a codec supported by the native libraries.
 // This is a one-of or union type: only one of these fields will be present.
-export type MessageContent = {
-  text?: string;
-  unknown?: UnknownContent;
-  reply?: ReplyContent;
-  reaction?: ReactionContent;
-  attachment?: StaticAttachmentContent;
-  remoteAttachment?: RemoteAttachmentContent;
-  readReceipt?: ReadReceiptContent;
-};
```

Comment:
This will be a breaking change for any integrator using this type directly should probably write something up about how to use this now in the future release notes.

Reply:
Hey just reading through & thought I would chime in here.

I think removing this really hurts the type story of the library. 

Couldn't something like [`AllowedCodecTypes` in this typescript playground](https://www.typescriptlang.org/play?ts=5.3.2#code/DYUwLgBADgTg9mOAuCBDAdgTwgXggbwF8AoYkADyjhkgEt0wQYAzVAYxAgAV5EBhOAxAMAKpiggAkgBMCxCAogB6AFQq0AVzAALarTDYA5nABuTdPUMQdtAM4Q2gxg2vjOKpfMWotumPswZFFswf3RDAG4vBVV1AwkIWmlhMFpmWiYID2jXCSCIELDInNiIAFtUACtqCDMYW1pBCDhma21OePdPRVqmBsEAWSrqFHQNMoAjJiie0rL6Grr+9GbWnQ63LO7FJcb0AYWYUfGpmCiSYiUlCAAlEFgQWxT7WAQ4ADoBIVE3GTJKaiQNjAVC2exfZxgMR5aQAHgAgr49AYZBAKM5pPZCpYADQQaFSWTo4SYgqhXEQABqfT2Q2qMDR5Ax9jGkyYeOp9Vph0ZzIgrNOAD4FPgvD4dMjAtIUIiJf4UdIvJ18gS-gpdoNhkcqTTNfSvBr9ocUJzlgd0NRSApHOhCho2IgYAAKOATSooUU9cV+AL5WU+hU5ZXS-G-RU9Q10kY6rl6y0R3VGi3a03c5NeQgASjkPRstne3slqLwrsqBaR8qlQe0dnewdwzTddbD1drka1DdL73b+tzNfzkZ5Jabg-TCguCkQAGVyeEnZngrOrJ7FDBwBoYCsAAYAEnwefLct90kISj3B+DhCQ5-73cTUZghHeN7b98OhC3GatZNQqTYEGYeAyhnIonRtSECRAyxFyKbMV2tQQQggABtJIoPCPFDXQwwAF0G3AlJIKXd5bCgYB9CdAByJBKMzHIbSQ5DC0rGQ8WDPC8DQ4jSPIsAqKUWj6MQyBkIqek8XmZMON6WN0GwkiyIoyj3kEno1zADcVnQEAAHcIAhQiwydeDvArY8cRyScwwsnp1XvLUUAAOROJgnTE6hMxs2zR2jZy2WdSSPK88c6PHb9bFQMoQHhWwnSSFADJ+GEETMhU8VVaQOXs8SYzNQ5BQXCAJjgOBQAwHNV3XTc2lrZjj1wHBOOkQ8AylCAADJ2pq-N60aprmxhL8Lk6CAABEQFYDRgDARKoTDBtZoy2FsQwskijxAV2X5FyYEFUgrlue41yeBgXl4D4AFF0EcZJpFm-4qBoRIhBYdhOCum6QDupwUlhRbNmJdBSXGybpv+mE8S4VAYEi8A+gbO5HBgOEVsMPFUcFYUKs6BKfqSkAvKgaHYcYeoUChmGotJ2wvFYYBgAmdgAGsAH4YMsLxHDKB4wT2Nntv8zm8bAFAAFV6DAAAOeEYBhzBiAuA6QfoThyolphWA4ACanWNFrrgaRLHKdga20mBsGYGpUAKCQ2DSWh-wIlxOkua5QVsOA7d-L6IB0-RtDQCBDFoMwVnBwlXYgeEHANkBHfKqYIDXYOQiYH2-Z0QO+HIlJrDgIrVa0OAKj-VB6ewegTDgJn0-adBI+0DBpHI8IY++SAWjaTgkeOqggaNp3IE6d4HsBZ7Sc1zgBGSNg7hT0JsBXA7HhbsAAFpDYiiZQDX7SmTXlvOAtNeNCeNepkttcN4mlW8QAAU6Ww2H8KB15XiWlGPihFLt9eMHlhCM8ABi1AwLCwJCgEGPgwbgLDIVWa0846wigVNGasCIZoCwMKAAPhADQQMb7aUVBcb+Y91avS1gg2ObA-roM4IDYGE1oFoPbhldKwoVyDwgfpOhXhhCfTAe3FAIg8TJzsAvXGM857iPNoVD6sdvrt1oawtwe0FAz1joIyEKB5G3VmsoiCqjREgHnubSRcdpGFEwIVEQtMy4M2ZlolIwjCqowgLg-ByR0hEIVvta45AyhgCgO8aghglB0wcWwJmrtI4iH7IkFkCBA5rjLm3SEuQCaJDoPYU+PtEAFFoIYdAqT8m6zXHbKAGQGCRx0L+LuaTc70HqVFMEqBDCcDsPUiJjMokQGSE-F+qQmhxVaDzFIdEDpNLYKCDo7Rmj+GDsU4ADTnabE6RaSAtgNBQEeowZqsT4mDwyRAAYospwiAgI5AA8hcxOuTZD5JOrIQeI9SFPQYpAcOQD7E9KZg2bSelw4yGMmKVKUoUCUQCUEkJMBDCUS8sGCF3TmbwoNNlaMABGLyPltQAAYLJZj8RAKFwTQlKEYEyGJB04mdKOSNTp9ztYMmtmRVATSKVAmFknI6jwUg+wmNga2DRuagDWhzN5nLbSfLoSIdE-zdI8JUTCEFCg6oKghSSmFcKEVhghRy1FdlZIPhQFitFRrjQQHxQrOio8nrCHGBAXRRsVxaGYJLBslFRYiCAWvSWqLhqbFlUyRBbBw4Nk6J3cOQawCkAlQ4EEYJ8TohDYkEVIAoqnUVZCENsJo0hvDujJcHCvDtMlYYiQ85hHJuoWGky6lNJZsMhIaNX4+H62SE4hg7NwhyPbV9fReaa10OLWpKqKwTJWQkLjJVIBo3BUUETSmcMyYED1jdaCjr22WHeK691hB50ISEfyBV0bdGuUzO8fhmjB6eRyCQMKXgNEdsHjovtijIS5urTPcOBVu3LiElKtdBsjZ4BeYukmfRL1bvCDkNIEAnRXsNq3TqQGkNWAAISNU3eu8IO6wBurgpZNo8A9IAsdbLUBW58HlLgEU2gAAvH2iGjZ7mY+ED8oVFD3sqhpaqZHo3jU+s6C9T6QCdrAO8G9raYjXHfgwDedhUDbxALvdEB8VYQAfm4AZtBX5rzk2AT+cAT5jCeNINeJhoY0wUMiqJ4m-2uKXO4vBBDvE+zrWOlzXiVbhggCQEhAJ3nCUbfjO4ZQEDRTAGAE2GbIB4DI0C6QKrNBHnVRAAARJq0J6WdUwhQOltc4XGAzl-A7eEUWYspBy2as0DkICmsNXlZMKArVZiiKQEaYWIvlei2wbQsX810PDW4SNdCuuMB65V6ptqh6bA+ubV+X0z3vtzngFcht2khD-TMAoZcRYQHFgwaWstUCYB2xaa6IAxYS2O3LHbTxn7gGu0dmWd2vBE0wMAOAqAQyHalq9075x2txs6+m7rFW+uxYbC6mAwBttC3bqNQpjx9uo3u3t57-2Ttna8BdjgmPbuA68A99SBOAc44UE-doUV4eHshAAGWEIYHQxx-I7fSKAYpNOxXhCB7GwLQIE32HG5F3r-WUgptoGm2L4JhY5pF5NyHEuh0zrxAriH4uGAjqDuAFZc0K2FXV2LgbKvy2cA87xsOY2wcTY17FnbE5UNiZfYdIrouptgF7Z9FbDBYRG494Nmd2uFBwadOhl5G5gAkWizQWwAB1f2Tp0vaCi1AWw6XMyEdssRuApGFUXQo86dLVP02cDKKfSAicU9BPT5mHbD7R2W4qrmNw06zf+6VwwA90BiZUz6B6IjdOUhI82-tl5g8R8o+75TjHBQ47BK0LQYA+YBWMFsCIOAAAJCg4mSJ7dvdnhQeOrtz7YAv1Iy-3ir8eBv7f5Bd-H4P4fknT3T-n6XyvzAa-b875eS-z30+BQfWpeKAf+wBUUgBg8jO4QLOEA2Eu+UBTOOgT+2eHOwgsMoBws7waBXOBMRG+6RGLu-G6IZ6wmUGAiEesOKBfmDu34om9m2GCi+iHemuaC36w6huNu7unekAFuDaE6eCsOowJ66IgmscZB9B4+ws1BQ+DAk+W2eu7w4Gfe9Qkmws8hYA3eEU00whekf2hOmAToyEOEMh-Igg+Ox6ehN25ORhJhWhccpOlhB21h2OthphJe3OYGvey6+YHheB2eiBMB2gKAi6TwkgDAu+yhPhah7c0BzO2geIGKuKphOBGBihUR1M2BS+6BEBd6tBdi9MvyDBLBsWjmRQzmnihC7mOQ9a1UW4fAGAgAmASQCbysrYDpZ7gvKpFRSEDpbvD4jxKoA7J9JwCPDoDNEFDbK7JoB27PDvBbi0GKzXBrwQDtDAASD1DEAjRThS6KTMCYC5rChrbIQADSz0EANcmAncIgOEwipxOEhAHUBAJAI05O4RhsHAtgsIF0oAUODC9gKS0gggwA2A+CTMFoOk6AxhRxxAAAkBdOQMCBoMkLCBTKkGXN8b8SkIKMhJRJzvEZRDhHiD8aXgwLifiToISYKBEBAPtGoPUkpqYJwJUBXmgNIKSO0mbA7PYKUnMoluPBrG9PIOoMKSlq1H6GCmqHELqqGINFkDJLVvSCaOijAKKTisqeamOCKfKXEsIGgDsjnDyfEqUjpPnJ9GwPYJfPUhFDXKsIyJFGRCAKKcKZ4JHEMLaeafYE0LrMCFUpAJyUwA7MQOQpPCcp7EzNnH6RALQjPPYP8SFmwUgigjAjOqxJgpgDidJMhIOjPGrlwYrqwSGjhIKJwtQrYOYhab4gcnqbrFuIPFuD3pTMcuynMk8k8Q8IupwG2cwPgg6HsPYI4FNLIInIzGZs0CsD6TnJmr+GLj7J6TrlyWwAcp0n7PTGgMvvnD9g8psBFMwOANgPkrkvUA4BXsXDHLGVsZsPCPTLnv2tQgSPYHgDsSKmkAcfgChO8RQOcW8f3J8bCAMOGZGSkLiZ6VSWgPYABVEkBWSZRKBThMhJ+eQPBbBXQoSbicGISbcdwN4dTP+YBVOWACBWWWhYhchYhiAFSchLinhIQHtCNAMI8BFO0rNA2G4rgteV9jpHeTPA+cQEAA) be done?

Usually the api user will know what codecs they want to use & so they could be instantiated as a param on the `Client` as in the js lib. That way the clients codecs could be generic and you could access all the information needed to  generate the content type that is needed for the `send` functions content for each codec.

It requires adding more type safety in a lot of places but worth the effort imo.

Reply:
> I think removing this really hurts the type story of the library.

Ok, you've convinced me. I'll take a look at bringing some more type safety around these changes.

Reply:
@peterferguson I brought in a similar approach as we use in the JS SDK here: https://github.com/xmtp/xmtp-js/pull/455.

Now `DecodedMessage#content()`'s type will be a union of possible types, based on codecs passed in at Client creation time.